Mathlib/Order/Prod/Lex/GroupWithZero.lean

```diff
@@ -0,0 +1,108 @@
+/-
+Copyright (c) 2025 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Algebra.Group.Prod
+import Mathlib.Algebra.Order.Hom.Monoid
+import Mathlib.Data.Prod.Lex
+
+/-!
+# Order homomorphisms for products of linearly ordered groups with zero
+
+This file defines order homomorphisms for products of linearly ordered groups with zero,
+which is identified with the `WithZero` of the lexicographic product of the units of the groups.
+
+The product of linearly ordered groups with zero `WithZero (αˣ ×ₗ βˣ)` is a
+linearly ordered group with zero itself with natural inclusions but only one projection.
+One has to work with the lexicographic product of the units `αˣ ×ₗ βˣ` since otherwise,
+the plain product `αˣ × βˣ` would not be linearly ordered.
+
+TODO: Create the "LinOrdCommGrpWithZero" category.
+
+-/
+
+namespace LinearOrderedCommGroupWithZero
+
+variable (α β : Type*) [LinearOrderedCommGroupWithZero α] [LinearOrderedCommGroupWithZero β]
+
+/-- Given linearly ordered groups with zero M, N, the natural inclusion ordered homomorphism from
+M to `WithZero (Mˣ ×ₗ Nˣ)`, which is the linearly ordered group with zero that can be identified
+as their product. -/
+@[simps!]
+def inl : α →*₀o WithZero (Lex (αˣ × βˣ)) where
```

Comment:
I'm struggling to understand why `WithZero (Lex (αˣ × βˣ))` is interesting; identifying the two zeros seems like a strange operation.

Reply:
If I have two different rank one valuations on a (higher local) field like Q_p((t)), then I can construct a new rank two valuation, mapping into the product of the codomains of the individual valuations, such that the rank one valuations commute with the projections.

Reply:
I guess my claim is that this map is actually hiding a composition of smaller maps; for instance, that `α →*₀o WithZero (αˣ)`, and that `α →*o β` lifts to `WithZero α →*o WithZero β`. Can you add these pieces?

Reply:
We have mul order isos between X and WithZero (Units X), as well as the equiv between mul isos between X and Y and WithZero X and WithZero Y (MulEquiv.withZero). Would you prefer this construction called out to those explicitly instead of directly constructing the terms?

Reply:
Yes, I think so; but also, the final construction seems very contrived and valuation centric, so perhaps it belongs in a file about valuations instead?

Reply:
You could imagine a category of linearly ordered comm groups with zero, and this would be the way that the (co)product would be constructed. And these are the homs.

Reply:
If one used those equivs, it becomes nastier because of the WithZero.rec term when casing on the nonzero branch of WithZero (Units X).

Reply:
> You could imagine a category of linearly ordered comm groups with zero, and this would be the way that the ~co~product would be constructed. And these are the homs.

Thanks, this is the interpretation I was missing. Do we have the version for the co-product of groups with zero? I think referencing that (after creating it) would make the picture clearer.